lightning/src/routing/router.rs

```diff
@@ -1449,17 +1456,24 @@ where L::Target: Logger {
 
 	// Draw multiple sufficient routes by randomly combining the selected paths.
 	let mut drawn_routes = Vec::new();
-	for i in 0..payment_paths.len() {
+	let mut prng = ChaCha20::new(random_seed_bytes, &[0u8; 12]);
+	let mut random_index_bytes = [0u8; ::core::mem::size_of::<usize>()];
+
+	let num_permutations = payment_paths.len();
+	for _ in 0..num_permutations {
 		let mut cur_route = Vec::<PaymentPath>::new();
 		let mut aggregate_route_value_msat = 0;
 
 		// Step (6).
-		// TODO: real random shuffle
-		// Currently just starts with i_th and goes up to i-1_th in a looped way.
-		let cur_payment_paths = [&payment_paths[i..], &payment_paths[..i]].concat();
+		// Do a Fisher-Yates shuffle to create a random permutation of the payment paths
+		for cur_index in (1..payment_paths.len()).rev() {
+			prng.process_in_place(&mut random_index_bytes);
+			let random_index = usize::from_be_bytes(random_index_bytes).wrapping_rem(cur_index+1);
```

Comment:
Could use `% (cur_index + 1)`?

Reply:
Ah, as discussed recently in #1360, I'd like to stick with 'functional style' math in cases like this.

Reply:
Fair enough though note the `wrapping_` variation is not needed.

Reply:
> Will need to insert the last commit between the first two commits given it's a fixup for the first commit and will be squashed into it whereas the second commit is independent. I typically use `git rebase -i` when needing to interleave fixups.

Ah, I now already squashed the new commit into the first one.

Reply:
Ah, I guess they are implemented the same. I don't have a strong opinion here, so feel free to leave it as is. I think the trait is mainly to allow operator overloading and is implemented by primitives for completeness.